zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -37,183 +38,187 @@
       .orElseThrow(RejectedExecutionException::new); // TODO: make an exception message
+    limiterMetrics.requests.increment();
+    limiterListener = limiterMetrics.wrap(limiterListener);
+
+    LimiterReleasingCallback releasingCallback =
+      new LimiterReleasingCallback(callback, isOverCapacity, limiterListener);
 
     try {
-      executor.execute(new QueuedCall<>(this, callback, limitListener));
-    } catch (RuntimeException | Error e) {
-      propagateIfFatal(e);
-      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
-      // write bound, but a drop in concurrency won't necessarily help.
-      limitListener.onIgnore();
-      throw e;
+      executor.execute(new EnqueueAndAwait(this, releasingCallback));
+    } catch (RuntimeException | Error t) { // possibly rejected, but from the executor, not storage!
+      propagateIfFatal(t);
+      callback.onError(t);
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle. Though
+      // we may still be write bound, but a drop in concurrency won't necessarily help.
+      limiterListener.onIgnore();
+      throw t;
     }
   }
 
-  @Override public Call<V> clone() {
-    return new ThrottledCall<>(executor, limiter, isOverCapacity, delegate.clone());
+  @Override public Call<Void> clone() {
+    return new ThrottledCall(delegate.clone(), executor, limiter, limiterMetrics, isOverCapacity);
   }
 
   @Override public String toString() {
     return "Throttled(" + delegate + ")";
   }
 
-  static String setCurrentThreadName(String name) {
-    Thread thread = Thread.currentThread();
-    String originalName = thread.getName();
-    thread.setName(name);
-    return originalName;
+  static final class AwaitableCallback implements Callback<Void> {
+    final CountDownLatch countDown = new CountDownLatch(1);
+    final AtomicReference<Throwable> throwable = new AtomicReference<>();
+
+    @Override public void onSuccess(Void ignored) {
+      countDown.countDown();
+    }
+
+    @Override public void onError(Throwable t) {
+      throwable.set(t);
+      countDown.countDown();
+    }
   }
 
-  static final class QueuedCall<V> implements Runnable {
-    final Call<V> delegate;
+  static final class EnqueueAndAwait implements Runnable {
+    final Call<Void> delegate;
     final Predicate<Throwable> isOverCapacity;
-    final Callback<V> callback;
-    final Listener limitListener;
+    final LimiterReleasingCallback callback;
 
-    QueuedCall(ThrottledCall<V> throttledCall, Callback<V> callback, Listener limitListener) {
+    EnqueueAndAwait(ThrottledCall throttledCall, LimiterReleasingCallback callback) {
       this.delegate = throttledCall.delegate;
       this.isOverCapacity = throttledCall.isOverCapacity;
       this.callback = callback;
-      this.limitListener = limitListener;
     }
 
+    /**
+     * This awaits callback completion in order to slow down (throttle) calls.
```

Comment:
It took me a while to follow how the throtlling works - to confirm, is it because the callbacks are run on a fixed number of threads? And the reason to use an `Executor` is to have both a limit on concurrent calls and control on the size of queued tasks, whereas a semaphore like the old ES implementation only handles concurrent? I think this makes sense just confirming my understanding.

Reply:
@Logic-32 can you validate this? I really do want to have the better docs here without getting too long winded. It isn't obvious at the moment.

Reply:
I'm sure I could go through the original PR and other supporting docs we made and find an answer for this.  But in the interest of not going cross-eyed, the short answer is that @anuraaga is correct.

A Semaphore is insufficient for the throttling goals as it would aggressively drop Spans.  The Executor offers us both the flexibility of rate-limiting requests to storage as well as the buffer (queue) so we don't have to aggressively drop Spans while storage is being sluggish.  It does all this in a nice, standardized bundle.

The control on the buffer size, btw, is so we don't reintroduce the original issue that the Semaphore was introduced to try and prevent.  Namely, OutOfMemoryErrors.  If a user desired, they could totally put an unbounded queue on the Executor but they'd be doing so at their own risk.

Reply:
Thanks for confirming! I think the words "slow down" confused me the most as I started looking for introducing artificial delay or something. Maybe just `waits until completion to ensure the number of executing calls doesn't surpass the concurrency limit of the executor`.